### ISSUE TITLE
Custom objectTypes setup causes invalid init of Realm instance

### ISSUE THREAD

**Issue by ppamorim** [realm/realm-cocoa#6205]
Creating a custom `Realm.Configuration` with determined types using `objectTypes ` parameter is causing future instances of `Realm.Configuration` to change/init with the last set object types.

Given this example:

The app contains these objects: User, Post, Comment, Photo

If you start the Realm in this way just to get User object, the code works fine:

```swift
  let configuration = Realm.Configuration(
      encryptionKey: RealmKeyManager.getRealmKey(), //Encryption key
      schemaVersion: SCHEMA_VERSION,
      migrationBlock: migrationBlock,
      objectTypes: [User.self])
  let realm = try Realm(configuration: configuration)
  realm.objects(User.self).first //Returns the first user.
```

So in another moment in the future (can be concurrent, I suspect) you decide to load all objects:

```swift
  let configuration = Realm.Configuration(
      encryptionKey: RealmKeyManager.getRealmKey(), //Encryption key
      schemaVersion: SCHEMA_VERSION,
      migrationBlock: migrationBlock)
  let realm = try Realm(configuration: configuration)
  realm.objects(Post.self).first //Post object contains User, Comment and Photo object
```

_This is invariant:_

```swift
  let configuration = Realm.Configuration(
      encryptionKey: RealmKeyManager.getRealmKey(), //Encryption key
      schemaVersion: SCHEMA_VERSION,
      migrationBlock: migrationBlock,
      objectTypes: [User.self, User.self, Comment.self, Photo.self]))
  let realm = try Realm(configuration: configuration)
  realm.objects(Post.self).first //Object contains User, Comment and Photo types
```

The two last examples above can randomly make `realm` instance not load the passed `objectTypes`. Could this be a race condition? I bet the first configuration could be loaded in a async queue (I am not really sure because the project is big and it will take a bit of time to investigate).

Some real reports (if you are blind sorry for the screenshots, but you can understand the code above,  nothing special here down below just some real life examples):

Given this script to load all possible objectTypes:

<img width="704" alt="Screenshot 2019-07-19 at 18 56 23" src="https://user-images.githubusercontent.com/5521682/61555313-01c3da00-aa57-11e9-941f-3f6070b84208.png">

Crash inside Realm core:

<img width="1207" alt="crash1" src="https://user-images.githubusercontent.com/5521682/61554798-b1984800-aa55-11e9-8ee9-b8b63a789915.png">

Logs from `createObject` function (lines 174 and 176 from the Swift source above):

<img width="1396" alt="crash2" src="https://user-images.githubusercontent.com/5521682/61555087-6fbbd180-aa56-11e9-8efe-ac3c5127448a.png">

This issue is totally related to issue #5777 .
I really want to replicate this in a sample app, I tried in the past but I wasn't able to success the reproduction. I hope that these information above help us.

## Version of Realm and Tooling

Realm framework version: 3.15.0

Realm Object Server version: No

Xcode version: Any (currently using 10.2.1).

iOS/OSX version: Any 

Dependency manager + version: Git submodule but this was happening in the past with Cocoapods.

**Comment by tgoyne** [comment]
If the Realm is already open on the current thread, `Realm()` will just hand you a reference to the existing object rather than reopening it. There's some validation to ensure that the configuration matches, but `objectTypes` currently isn't one of the properties checked.

This is something we should probably fix, but what are you actually trying to achieve by sometimes opening it with a subset of types and sometimes with all types? It isn't obvious to me how that could ever be a useful thing to do.

**Comment by ppamorim** [comment]
@tgoyne Should I keep trying to load the Realm instance until it loads the desired `objectTypes`? I think these Realm instances are loaded in the main thread.

**Comment by ppamorim** [comment]
@tgoyne The issue is still happening on 3.17.1.

**Comment by tgoyne** [comment]
No, you should not try to keep opening the Realm until it happens to work. You should just not be trying to open the Realm with inconsistent configurations to begin with.

**Comment by ppamorim** [comment]
@tgoyne Any idea of workaround that I could implement? Is this problem from Realm-core?

**Comment by tgoyne** [comment]
Why are you trying to open the Realm with inconsistent configurations?

**Comment by ppamorim** [comment]
@tgoyne I read somewhere that dynamically loading a Realm instance with the only needed data helps to keep backgrouding initialized from `UIApplication.shared.beginBackgroundTask` running for longer in order to keep the memory pressure lower, so the system doesn't kill the app.

**Comment by tgoyne** [comment]
When you don't explicitly specify the objectTypes in a Realm, we have to ask the runtime for a list of all of the classes present in the application and filter it, which causes a one-time spike in memory usage (both from getting the list of classes and because it eagerly loads delay-loaded frameworks). Only passing in a subset of your classes to objectTypes doesn't do much of anything for memory usage.

**Comment by ppamorim** [comment]
@tgoyne What about allocation?

**Comment by tgoyne** [comment]
What about it?

**Comment by ppamorim** [comment]
So far I know, Realm allocates the same amount of bytes in the memory for the Realm file, so just loading a single objectType should allocate the amount needed only to fulfill all the data from this objectType. 

What about the situation reported at https://realm.io/docs/swift/latest/#class-subsets? The problem could happen in there.

**Comment by tgoyne** [comment]
The types you pass in for objectTypes does not have any impact on what data is read from the file when opening a pre-existing Realm file, and we always fully initialize our internal schema data structures for all of the types present in the file (and then only expose the ones specifically requested).

That section of the docs is intending to be referring to different components using different Realm files; separate components trying to share a single Realm file with different class subsets won't work very well.

**Comment by ppamorim** [comment]
@tgoyne Well thank you for the explanation, I will change it on our app.
I think the documentation should include this characteristic (or maybe be patched?).

**Comment by ppamorim** [comment]
Basically my issue came from here:

<img width="891" alt="Screenshot 2019-08-11 at 14 53 40" src="https://user-images.githubusercontent.com/5521682/62834686-ee67f100-bc47-11e9-9bdb-156900979cc3.png">

**Comment by Splash04** [comment]
130 classes => crash on iPhones with 2 GB of RAM (iPhone 7, iPhone SE 1st, iPod Touch 7):
`malloc_ptr classes(objc_copyClassList(&numClasses), &free);`

Similar issue here:
https://github.com/realm/realm-swift/issues/8026

So need to provide list manually:

     If you would like to get a currently up-to-date list of classes:
     1. Set Realm.Configuration(objectTypes: nil) for automatic calculation
     2. Set break point on Realm.RLMSchema at line:
     [s_localNameToClass enumerateKeysAndObjectsUsingBlock:^(NSString *, Class cls, BOOL *) {
     3. Run on simulator or device with large Ram
     4. Print description of s_localNameToClass value
     
     Additionally, you can:
     5. Copy list to Sublime text
     6. Press Command+A to select all lines
     7. Press Command+Shift+L to start editing all of them
     8. Command + Left to move at the start of the line
     9. Option + Right to move after first word
    10. Shift+Command+Right to select all text after first word
     11. Type ".self," to add needed data
     12. Press Esc to end multiline editing
     13. Rename RealmSwiftDynamicObject to DynamicObject
     14. Copy list to "objectTypes" array
     15. Make sure that array will be set (not nil):  Realm.Configuration(objectTypes: CREATED_ARRAY) 
     
And remember, that after this code changes, you will need to add each new realm class to the list. In other way, you will receive crash:
`Invalid keypath 'XXXX: Property 'XXXX' not found in object of type 'ApiObject'`
